pkg/namespace/namespace.go

```diff
@@ -0,0 +1,130 @@
+package namespace
+
+import (
+	"bytes"
+	"fmt"
+)
+
+type Namespace struct {
+	Version uint8
+	ID      []byte
+}
+
+// New returns a new namespace with the provided version and id.
+func New(version uint8, id []byte) (Namespace, error) {
+	err := validateVersion(version)
+	if err != nil {
+		return Namespace{}, err
+	}
+
+	err = validateID(version, id)
+	if err != nil {
+		return Namespace{}, err
+	}
+
+	return Namespace{
+		Version: version,
+		ID:      id,
+	}, nil
+}
+
+// MustNew returns a new namespace with the provided version and id. It panics
+// if the provided version or id are not supported.
+func MustNew(version uint8, id []byte) Namespace {
+	ns, err := New(version, id)
+	if err != nil {
+		panic(err)
+	}
+	return ns
+}
+
+// MustNewV0 returns a new namespace with version 0 and the provided id. This
+// function panics if the provided id is not exactly 10 bytes.
+func MustNewV0(id []byte) Namespace {
+	if len(id) != NamespaceVersionZeroIDSize {
+		panic(fmt.Sprintf("invalid namespace id length: %v must be %v", len(id), NamespaceVersionZeroIDSize))
+	}
+
+	ns, err := New(NamespaceVersionZero, append(NamespaceVersionZeroPrefix, id...))
+	if err != nil {
+		panic(err)
+	}
+	return ns
+}
+
+// From returns a namespace from the provided byte slice.
+func From(b []byte) (Namespace, error) {
+	if len(b) != NamespaceSize {
+		return Namespace{}, fmt.Errorf("invalid namespace length: %v must be %v", len(b), NamespaceSize)
+	}
+	rawVersion := b[0]
+	rawNamespace := b[1:]
+	return New(rawVersion, rawNamespace)
+}
+
+// Bytes returns this namespace as a byte slice.
+func (n Namespace) Bytes() []byte {
+	return append([]byte{n.Version}, n.ID...)
+}
+
+// ValidateBlobNamespace returns an error if this namespace is not a valid blob namespace.
+func (n Namespace) ValidateBlobNamespace() error {
+	if n.IsReserved() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use a reserved namespace ID, want > %v", n.Bytes(), MaxReservedNamespace.Bytes())
+	}
+
+	if n.IsParityShares() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use parity shares namespace ID", n.Bytes())
+	}
+
+	if n.IsTailPadding() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use tail padding namespace ID", n.Bytes())
+	}
+
+	return nil
+}
```

Comment:
shouldn't we also be checking that only a portion of the namespace is used? is that in a different PR?

Reply:
thanks, I think I was just confused by why we had a separation of validation logic, but

> I don't think it makes sense to enforce this validation on all usage of New because all reserved namespaces will fail the ValidateBlobNamespace check

explains it